### PR TITLE
[SPARK-28216][SQL][TEST] Add `getLocalDirSize` to SQLTestUtils

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This pr add calculate local directory size to `SQLTestUtils`.

We can avoid these changes after this pr:
image

## How was this patch tested?

Existing test

### CODE REVIEW

**Comment by wangyum** [comment]
cc @HyukjinKwon @gengliangwang

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala [hunk 1]
```diff
@@ -424,6 +425,14 @@ private[sql] trait SQLTestUtilsBase
   protected def testFile(fileName: String): String = {
     Thread.currentThread().getContextClassLoader.getResource(fileName).toString
   }
+
+  /**
+   * Returns the size of the local directory except the metadata file and the temporary file.
+   */
+  def getLocalDirSize(file: File): Long = {
+    assert(file.isDirectory)
+    file.listFiles.filter(f => DataSourceUtils.isDataFile(f.getName)).map(_.length).sum
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: Nit: we can use `f.toPath` so that we don't need to modify `DataSourceUtils`

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala [hunk 2]
```diff
@@ -295,9 +295,6 @@ abstract class StatisticsCollectionTestBase extends QueryTest with SQLTestUtils
     }
   }
 
-  def getDataSize(file: File): Long =
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: Why this change?

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: Move `StatisticsCollectionTestBase.getDataSize` to `SQLTestUtils.getLocalDirSize`.

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: It's `java.nio.file.Path`, not `org.apache.hadoop.fs.Path`:
image

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The newly added function `getLocalDirSize` is a different one. Since the function semantic is changed, we can say `getDataSize` is technically removed. As a result of semantic changes, the following is introduced in the test case by this PR.
```scala
+ val tableLocationSize = new File(new URI(tableLocation)).listFiles()
+   .filter(!_.getName.endsWith(".crc")).map(_.length()).sum
```
Please update the PR title and description.

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: But I can see the same code in StatisticsCollectionSuite.scala

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionTestBase.scala]: Let's not remove this function to avoid confusion, @wangyum .

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 3]
```diff
@@ -642,10 +642,12 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
 
           val stagingFileSize = staging.length()
           val metadataFileSize = metadata.length()
-          val tableLocationSize = getDataSize(new File(new URI(tableLocation)))
+          val tableLocationSize = new File(new URI(tableLocation)).listFiles()
+            .filter(!_.getName.endsWith(".crc")).map(_.length()).sum
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: Let's use the existing `getDataSize`.

**Comment by dongjoon-hyun** [comment]
Hi, @wangyum .
The contribution of `getLocalDirSize` is clear and enough. Please keep the old `getDataSize` because it works differently. Using new function instead of `getDirSize` is okay.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @wangyum !

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you, @wangyum . This is useful.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 4]
```diff
@@ -60,8 +60,8 @@ object DataSourceUtils {
...
   // counted as data files, so that they shouldn't participate in tasks like
   // location size calculation.
-  private[sql] def isDataPath(path: Path): Boolean = {
-    val name = path.getName
-    !(name.startsWith("_") || name.startsWith("."))
-  }
+  private[sql] def isDataPath(path: Path): Boolean = isDataFile(path.getName)
+
+  private[sql] def isDataFile(fileName: String) =
+    !(fileName.startsWith("_") || fileName.startsWith("."))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: This is actually different from our code which is calling `fs.getContentSummary(tablePath).getLength` in https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala#L125 

getContentSummary is using isChecksumFile to filter out the metadata file. 

https://hadoop.apache.org/docs/r2.7.4/api/src-html/org/apache/hadoop/fs/ChecksumFileSystem.html

```Java
093  /** Return true iff file is a checksum file name.*/
094  public static boolean isChecksumFile(Path file) {
095    String name = file.getName();
096    return name.startsWith(".") && name.endsWith(".crc");
097  }
```

Obviously, they are different.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 5]
```diff
@@ -121,7 +121,7 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
     withTempDir { tempDir =>
       withTable("t1") {
         spark.range(5).write.mode(SaveMode.Overwrite).parquet(tempDir.getCanonicalPath)
-        val dataSize = getDataSize(tempDir)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: At least, we need to revert this change.

Comment (gatorsmile) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Also correct the impl of `getDataSize`?

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: OK. will do it later.
